### PR TITLE
Fix frequent agent drain mode alerts

### LINKED ISSUE #11976: Frequent AgentDrainMode Alerts
**Impact of the bug**
WMAgent

**Describe the bug**
WMAgent has been frequently sending email alerts with the following message:
```Agent had a drain status transition to AgentDrainMode = False.```

When the disk usage for /data exceeds 85%, the agents automatically set AgentDrainMode to true. This places the agent in drain mode, preventing it from accepting new workflows, and an email alert is sent.
As workflows in the agent clear out and get archived the disk usage falls and the agent then sets `AgentDrainMode` to false causing this cycle to continue.

**How to reproduce it**
Described above.

**Expected behavior**
Based on Alan's suggestion in the email thread, if the agent is in `UserDrainMode`, we enable `AgentDrainMode` (if disk usage is above the threshold) and never switch it back to False.

### PR DESCRIPTION
**Pull request by hassan11196**
Fixes #11976

#### Status
<In development | not-tested | on hold | ready>

#### Description
Fixes Frequent AgentDrainMode Alerts 

#### Is it backward compatible (if not, which system it affects?)
YES

#### Related PRs
None

#### External dependencies / deployment changes
None

### CODE REVIEW

File: src/python/WMCore/Services/ReqMgrAux/ReqMgrAux.py [hunk 1]
```diff
@@ -410,8 +410,20 @@ def listDiskUsageOverThreshold(config, updateDB):
 
     if updateDB and not t0Flag:
         agentDrainMode = bool(len(overThresholdDisks))
+        updateWMConfig = False
```
Comment (amaltaro) [inline src/python/WMCore/Services/ReqMgrAux/ReqMgrAux.py:413]: Ahmed, if what we are trying to accomplish is to simply not set AgentDrainMode if UserDrainMode is already True, then I would suggest a potentially simpler variation for this. Something like:
```
agentDrainMode = bool(len(overThresholdDisks))
if agentConfig and (agentDrainMode != agentConfig["AgentDrainMode"]):
    if agentConfig["UserDrainMode"] == 'True':
        print("Agent is already in UserDrainMode: True mode. No need to update.")
    else:
        reqMgrAux.updateWMAgentConfig(config.Agent.hostName, {"AgentDrainMode": agentDrainMode},
                                          inPlace=True)
```

File: src/python/WMCore/Services/ReqMgrAux/ReqMgrAux.py [hunk 2]
```diff
@@ -410,8 +410,20 @@ def listDiskUsageOverThreshold(config, updateDB):
...
         if agentConfig and (agentDrainMode != agentConfig["AgentDrainMode"]):
-            reqMgrAux.updateWMAgentConfig(config.Agent.hostName, {"AgentDrainMode": agentDrainMode},
+            
+            if agentConfig["UserDrainMode"] == 'True':
+                if agentDrainMode == 'False':
+                    print("Agent is already in UserDrainMode: True mode. No need to update.")
```
Comment (amaltaro) [inline src/python/WMCore/Services/ReqMgrAux/ReqMgrAux.py:418]: Can you please replace the print statement by the logging module (I'd suggest info level)?

**Review by amaltaro — CHANGES_REQUESTED** [review]
@hassan11196 thank you for providing this fix. I left a couple of comments along the code for your consideration.

File: src/python/WMCore/Services/ReqMgrAux/ReqMgrAux.py (hunk 1 above)
Comment (hassan11196) [reply src/python/WMCore/Services/ReqMgrAux/ReqMgrAux.py]: updated Thank you Alan

File: src/python/WMCore/Services/ReqMgrAux/ReqMgrAux.py (hunk 2 above)
Comment (hassan11196) [reply src/python/WMCore/Services/ReqMgrAux/ReqMgrAux.py]: modified to use logging at `INFO` level.

File: src/python/WMCore/Services/ReqMgrAux/ReqMgrAux.py [hunk 3]
```diff
@@ -411,7 +411,11 @@ def listDiskUsageOverThreshold(config, updateDB):
     if updateDB and not t0Flag:
         agentDrainMode = bool(len(overThresholdDisks))
         if agentConfig and (agentDrainMode != agentConfig["AgentDrainMode"]):
-            reqMgrAux.updateWMAgentConfig(config.Agent.hostName, {"AgentDrainMode": agentDrainMode},
+            if agentConfig["UserDrainMode"] == 'True':
```
Comment (amaltaro) [inline src/python/WMCore/Services/ReqMgrAux/ReqMgrAux.py:414]: Looking into this again, it looks like these drain metrics are actually of the boolean type, e.g.:
https://cmsweb.cern.ch/reqmgr2/data/wmagentconfig/cmsgwms-submit13.fnal.gov

so I guess we should update this if statement to `if agentConfig["UserDrainMode"] is True:` or `if agentConfig["UserDrainMode"]:`

Perhaps you can check it in that testbed agent that you can play with (do you still have such agent?)?

**Review by amaltaro — CHANGES_REQUESTED** [review]
Just proceeding with the formal "review" to make sure it goes back to you.

File: src/python/WMCore/Services/ReqMgrAux/ReqMgrAux.py (hunk 3 above)
Comment (hassan11196) [reply src/python/WMCore/Services/ReqMgrAux/ReqMgrAux.py]: Hi Alan, 
Yes I have the vocms264 testbed agent.

I tested out the Pull Request and verified that it's working. I have attached the document with logs and issues I encountered, Please check it out and let me know if everything looks good.

Thank you.

[Hassan_PR_12142_Agent_Drain_Mode_Alert_Tesst.pdf](https://github.com/user-attachments/files/17377904/Hassan_PR_12142_Agent_Drain_Mode_Alert_Tesst.pdf)

**Comment by amaltaro** [comment]
@hassan11196 these changes are looking good to me know. Can you please squash all these commits into a single one? In case you need some assistance with that, please check out Step 10 in this doc: https://github.com/dmwm/WMCore/blob/master/CONTRIBUTING.rst

**Comment by hassan11196** [comment]
> @hassan11196 these changes are looking good to me know. Can you please squash all these commits into a single one? In case you need some assistance with that, please check out Step 10 in this doc: https://github.com/dmwm/WMCore/blob/master/CONTRIBUTING.rst

Hi @amaltaro, I have squashed the commits into one. Let me know if its okay.
Thanks

File: src/python/WMCore/Services/ReqMgrAux/ReqMgrAux.py [hunk 4]
```diff
@@ -410,8 +410,12 @@ def listDiskUsageOverThreshold(config, updateDB):
 
     if updateDB and not t0Flag:
         agentDrainMode = bool(len(overThresholdDisks))
-        if agentConfig and (agentDrainMode != agentConfig["AgentDrainMode"]):
-            reqMgrAux.updateWMAgentConfig(config.Agent.hostName, {"AgentDrainMode": agentDrainMode},
+        if agentConfig and (agentDrainMode != agentConfig.get("AgentDrainMode")):
+            if agentConfig.get("UserDrainMode") is True:
+                    logging.info("Agent is already in UserDrainMode: True mode. No need to update.")
```
Comment (amaltaro) [inline src/python/WMCore/Services/ReqMgrAux/ReqMgrAux.py:415]: It looks like you got wrong indentation in this line. As python is picky with those, can you please correct it?
The jenkins report (at the bottom, pycodestyle) is giving you hints of where those are.

**Review by amaltaro — APPROVED** [review]

**Comment by amaltaro** [comment]
Thank you for the follow up, Ahmed!

**Pull request merged**